### PR TITLE
feat: add Kreuzberg document converter integration

### PR DESCRIPTION
**Pull request by v-tan**
### Related Issues

- part of #2926

### Proposed Changes:

Adds a `KreuzbergConverter` component under `integrations/kreuzberg/`.

**Component design:**

- Implements the `@component` protocol with `sources` and `meta` inputs, `documents` and `raw_extraction` outputs
- Delegates extraction to kreuzberg's sync APIs (`extract_file_sync`, `batch_extract_files_sync`) — async pipeline support is deferred
- `ByteStream` inputs are written to a temp file before extraction (kreuzberg requires file paths)
- Batch mode (default) uses kreuzberg's Rust rayon thread pool for parallel extraction; can be disabled for single-file fallback
- Three extraction modes: whole-document (default), per-page (`PageConfig`), and chunked (`ChunkingConfig`) — each produces different Document granularity and metadata shape

**Serialization:**

- `ExtractionConfig` is serialized via kreuzberg's `config_to_json` / `config_merge` utilities
- `config_path` stored as POSIX string for cross-platform compatibility
- `easyocr_kwargs` serialized as plain dict
- Round-trip tested via `to_dict` / `from_dict`

**Metadata handling:**

- Kreuzberg's `ExtractionResult` fields are flattened into `Document.meta`
- Overlap keys (`quality_score`, `output_format`, `keywords`) are handled from top-level fields to avoid duplication
- User-provided `meta` is deep-copied to prevent mutation across documents
- Format-specific metadata (e.g. PDF title, author, page count) is included when available

**Files added:**

```
integrations/kreuzberg/
├── src/haystack_integrations/components/converters/kreuzberg/
│   ├── __init__.py
│   └── converter.py          # ~800 lines (component + helpers)
├── tests/
│   ├── test_converter.py     # ~1200 lines (unit tests)
│   └── test_converter_integration.py  # ~600 lines (integration tests)
├── pyproject.toml
├── pydoc/config_docusaurus.yml
├── README.md
└── LICENSE.txt
.github/workflows/kreuzberg.yml   # CI: nightly + PR triggers
.github/labeler.yml               # label entry
README.md                         # inventory table entry
```

### How did you test it?

- `hatch run test:unit` — serialization round-trips, metadata extraction, error handling, edge cases (empty files, unsupported formats, blank pages), batch vs single-file modes
- `hatch run test:integration` — end-to-end extraction with real PDF, DOCX, HTML, TXT fixtures; pipeline integration with `DocumentCleaner` and `DocumentWriter`
- No external services required — kreuzberg processes locally. OCR tests require Tesseract installed.

### Notes for the reviewer

- `converter.py` is ~800 lines because it handles three extraction modes, metadata flattening from kreuzberg's rich result types, and serialization of kreuzberg's config objects — open to splitting if preferred
- The `raw_extraction` output exposes serialized `ExtractionResult` for debugging without requiring kreuzberg as a dependency downstream
- kreuzberg's error handling uses error codes (`get_last_error_code`, `get_error_details`) — the converter logs warnings and skips failed files rather than raising

### Checklist

- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- [x] I have updated the related issue with new insights and changes
- [x] I added unit tests and updated the docstrings
- [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

File: .github/workflows/kreuzberg.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+          hatch -e test env run -- uv pip install git+https://github.com/deepset-ai/haystack.git@main
+          hatch run test:unit
+
+      - name: Send event to Datadog for nightly failures
+        if: failure() && github.event_name == 'schedule'
+        uses: ./.github/actions/send_failure
+        with:
+          title: |
+            Core integrations nightly tests failure: ${{ github.workflow }}
+          api-key: ${{ secrets.CORE_DATADOG_API_KEY }}
```
Comment (anakin87) [inline .github/workflows/kreuzberg.yml]:
```suggestion

  notify-slack-on-failure:
    needs: run
    if: failure() && github.event_name == 'schedule'
    runs-on: ubuntu-slim
    steps:
      - uses: deepset-ai/notify-slack-action@v1
        with:
          slack-webhook-url: ${{ secrets.SLACK_WEBHOOK_URL_NOTIFICATIONS }}
```

we recently switched to this new approach

File: .github/workflows/kreuzberg.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+      - uses: actions/checkout@v6
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v6
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install Hatch
+        run: pip install hatch "virtualenv<21.0.0"
```
Comment (anakin87) [inline .github/workflows/kreuzberg.yml]:
```suggestion
        run: pip install hatch
```

this incompatibility has been fixed

File: integrations/kreuzberg/README.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
+# kreuzberg-haystack
+
+[![PyPI - Version](https://img.shields.io/pypi/v/kreuzberg-haystack.svg)](https://pypi.org/project/kreuzberg-haystack)
+[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/kreuzberg-haystack.svg)](https://pypi.org/project/kreuzberg-haystack)
+
+- [Changelog](https://github.com/deepset-ai/haystack-core-integrations/blob/main/integrations/kreuzberg/CHANGELOG.md)
+
```
Comment (anakin87) [inline integrations/kreuzberg/README.md:7]: Please stick to [this format](https://github.com/deepset-ai/haystack-core-integrations/blob/main/integrations/chroma/README.md) and remove everything else.

We'll put these useful examples in documentation page + integration page

File: integrations/kreuzberg/kreuzberg.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
+---
```
Comment (anakin87) [inline integrations/kreuzberg/kreuzberg.md]: this file is meant to be automatically generated when an integration is released. Please remove it from this PR

File: integrations/kreuzberg/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+[tool.mypy]
+strict = true
+install_types = true
+non_interactive = true
+
+[tool.pytest.ini_options]
+addopts = "--strict-markers"
+markers = [
+    "unit: unit tests",
```
Comment (anakin87) [inline integrations/kreuzberg/pyproject.toml]:
```suggestion
```

we don't explicitly mark unit tests

File: integrations/kreuzberg/README.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+---
+
+## Contributing
+
+Refer to the general [Contribution Guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md).
+
+No external services or API keys are required — kreuzberg processes everything locally. For OCR tests, ensure Tesseract is installed on your system.
```
Comment (anakin87) [inline integrations/kreuzberg/README.md]:
```suggestion
For OCR tests, ensure Tesseract is installed on your system.
```

File: integrations/kreuzberg/uv.lock [hunk 7]
```diff
@@ -0,0 +1,1184 @@
+version = 1
```
Comment (anakin87) [inline integrations/kreuzberg/uv.lock]: please remove this file

File: integrations/kreuzberg/tests/test_files/sample.html [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
```
Comment (anakin87) [inline integrations/kreuzberg/tests/test_files/sample.html:1]: nit: for consistency with other integrations, please move all test files to `integrations/kreuzberg/tests/test_files`

File: integrations/kreuzberg/tests/test_converter.py [hunk 9]
```diff
@@ -0,0 +1,1191 @@
...
+    _serialize_page_tables,
+    _serialize_tables,
+    _serialize_warnings,
+)
+
+FIXTURES_DIR = Path(__file__).parent / "fixtures"
+CONVERTER_MODULE = "haystack_integrations.components.converters.kreuzberg.converter"
+
+
+@pytest.mark.unit
```
Comment (anakin87) [inline integrations/kreuzberg/tests/test_converter.py]: we don't usually mark unit tests. Please remove `@pytest.mark.unit` markers

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 10]
```diff
@@ -0,0 +1,801 @@
...
+from typing import Any, cast
+
+from haystack import Document, component, default_from_dict, default_to_dict, logging
+from haystack.components.converters.utils import (
+    get_bytestream_from_source,
+    normalize_metadata,
+)
+from haystack.dataclasses import ByteStream
+
+from kreuzberg import (  # type: ignore[attr-defined]
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: Could you explain what the problem with mypy is here?

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 11]
```diff
@@ -0,0 +1,801 @@
...
+    )
+    ```
+
+    The converter exposes two output sockets: `documents` and
+    `raw_extraction`. The `raw_extraction` output contains the serialized
+    kreuzberg `ExtractionResult` for each source, useful for debugging or
+    advanced downstream processing.
+
+    **Token reduction** can be configured via
+    ``ExtractionConfig(token_reduction=TokenReductionConfig(mode="moderate"))``
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]:
```suggestion
    `ExtractionConfig(token_reduction=TokenReductionConfig(mode="moderate"))`
```

Let's use single backticks everywhere

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 12]
```diff
@@ -0,0 +1,801 @@
...
+    ) -> None:
+        """
+        Create a `KreuzbergConverter` component.
+
+        :param config:
+            An optional `kreuzberg.ExtractionConfig` object to customize
+            extraction behavior. Use this to set output format, OCR backend
+            and language, force-OCR mode, per-page extraction, chunking,
+            keyword extraction, and other kreuzberg options. If not provided,
+            kreuzberg's defaults are used.
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:109]: Please link kreuzberg documentation here in docstrings, so user would find this info in the (auto-generated) API reference

File: integrations/kreuzberg/tests/test_converter.py [hunk 13]
```diff
@@ -0,0 +1,1191 @@
+# SPDX-FileCopyrightText: 2026-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+import json
```
Comment (anakin87) [inline integrations/kreuzberg/tests/test_converter.py:4]: This might sound strange, but keeping the current high coverage, can we reduce a bit the number of tests in this module? This would help with review and future maintenance.

File: integrations/kreuzberg/tests/test_converter_integration.py [hunk 14]
```diff
@@ -0,0 +1,607 @@
+# SPDX-FileCopyrightText: 2026-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Integration tests for KreuzbergConverter.
+
+These tests call kreuzberg's real extraction APIs against fixture files —
+no mocking. Since kreuzberg is local-only (no external API), these do
+not require any environment variables or skip conditions.
+"""
```
Comment (anakin87) [inline integrations/kreuzberg/tests/test_converter_integration.py]:
```suggestion
```

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 15]
```diff
@@ -0,0 +1,801 @@
...
+            If `True`, the full file path is stored in the Document metadata.
+            If `False`, only the file name is stored.
+        :param batch:
+            If `True`, use kreuzberg's batch extraction APIs, which leverage
+            Rust's rayon thread pool for parallel processing. If `False`,
+            sources are extracted one at a time.
+        :param easyocr_kwargs:
+            Optional keyword arguments to pass to EasyOCR when using the
+            `"easyocr"` backend. Supports GPU, beam width, model storage,
+            and other EasyOCR-specific options.
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:125]: Link docs where users can find supported kwargs

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 16]
```diff
@@ -0,0 +1,801 @@
...
+    def from_dict(cls, data: dict[str, Any]) -> "KreuzbergConverter":
+        """
+        Deserialize this component from a dictionary.
+
+        :param data:
+            Dictionary to deserialize from.
+        :returns:
+            Deserialized component.
+        """
+        data = {**data, "init_parameters": dict(data.get("init_parameters", {}))}
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: I'm not sure if this line is needed.
I would expect `init_parameters` to always be present. If not present, deserialization must fail.
Agree?

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 17]
```diff
@@ -0,0 +1,801 @@
...
+        :param config:
+            An optional `kreuzberg.ExtractionConfig` object to customize
+            extraction behavior. Use this to set output format, OCR backend
+            and language, force-OCR mode, per-page extraction, chunking,
+            keyword extraction, and other kreuzberg options. If not provided,
+            kreuzberg's defaults are used.
+        :param config_path:
+            Path to a kreuzberg configuration file (`.toml`, `.yaml`, or
+            `.json`). When both `config` and `config_path` are provided,
+            `config` takes precedence (is merged on top of the file config).
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: I see that we both allow `config` and `config_path` and then we merge them in a complex way.
My impression is that just raising an error if both are provided would help simplify the code.
What do you think?

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 18]
```diff
@@ -0,0 +1,801 @@
...
+        """
+        # Expand directories
+        has_dirs = any(isinstance(s, (str, Path)) and Path(s).is_dir() for s in sources)
+        if has_dirs and isinstance(meta, list):
+            msg = (
+                "When directories are present in 'sources', 'meta' must be a "
+                "single dictionary, not a list, since the number of files in "
+                "a directory is not known in advance."
+            )
+            raise ValueError(msg)
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: nit: let's use `TypeError`

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 19]
```diff
@@ -0,0 +1,801 @@
...
+
+    @staticmethod
+    def supported_ocr_backends() -> list[str]:
+        """
+        List all OCR backends registered in kreuzberg.
+
+        :returns:
+            List of OCR backend names.
+        """
+        return cast(list[str], list_ocr_backends())
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: these do not seem to be used

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 20]
```diff
@@ -0,0 +1,801 @@
...
+
+
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: Some of the functions below could be moved to a utils.py file for clarity

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 21]
```diff
@@ -0,0 +1,779 @@
...
+        if isinstance(config_data, str):
+            init_params["config"] = _config_from_json_str(config_data)
+        return default_from_dict(cls, data)
+
+    def _build_config(self) -> ExtractionConfig:
+        """
+        Build the effective `ExtractionConfig` by merging all configuration
+        sources in priority order: file config < explicit config.
+
+        Always returns a *fresh* config object — never mutates ``self.config``.
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: please find all uses of \`\`code\`\` and replace them with \`code\`

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py (hunk 17 above)
Comment (v-tan) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: Agree. Resolved it by raising an error on presence of both params.

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py (hunk 10 above)
Comment (v-tan) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: kreuzberg ships `py.typed` but the `.pyi` stub for its Rust bindings is not included in the built wheel. It exists in the repo but isn't packaged. 
Without it, `mypy` cannot resolve any of the symbols we import. 
The suppress is required until this packaging issue is fixed upstream.

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py (hunk 16 above)
Comment (v-tan) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: Agree. `from_dict()` would raise error. removed it.

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py (hunk 10 above)
Comment (anakin87) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: let's handle this in pyproject for consistency with other integrations

https://github.com/deepset-ai/haystack-core-integrations/blob/1f8921172fdfdc11f1fa76a8626ec5537d317a0d/integrations/amazon_bedrock/pyproject.toml#L84-L91

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 22]
```diff
@@ -0,0 +1,518 @@
...
+        """
+        Create a `KreuzbergConverter` component.
+
+        :param config:
+            An optional `kreuzberg.ExtractionConfig` object to customize
+            extraction behavior. Use this to set output format, OCR backend
+            and language, force-OCR mode, per-page extraction, chunking,
+            keyword extraction, and other kreuzberg options. If not provided,
+            kreuzberg's defaults are used.
+            See the `kreuzberg API reference <https://docs.kreuzberg.dev/reference/api-python/>`_
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]:
```suggestion
            See the [kreuzberg API reference](https://docs.kreuzberg.dev/reference/api-python/>`)
```
Let's use this syntax everywhere

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 23]
```diff
@@ -0,0 +1,518 @@
...
+                        easyocr_kwargs=self.easyocr_kwargs,
+                    )
+                else:
+                    result = extract_file_sync(
+                        source,
+                        config=config,
+                        easyocr_kwargs=self.easyocr_kwargs,
+                    )
+                results.append(result)
+            except Exception as e:
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:526]: just a question. Is this behavior consistent with batch extraction? If we get conversion errors, are they just reported as warnings also in batch mode?

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 24]
```diff
@@ -0,0 +1,518 @@
...
+        if bytes_data:
+            bytes_results = batch_extract_bytes_sync(
+                bytes_data, bytes_mimes, config=config, easyocr_kwargs=self.easyocr_kwargs
+            )
+            for idx, result in zip(bytes_indices, bytes_results, strict=True):
+                results[idx] = result
+
+        return results
+
+    def _build_extraction_metadata(self, result: ExtractionResult) -> dict[str, Any]:
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: this seems like a static method

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 25]
```diff
@@ -0,0 +1,518 @@
...
+            if page_images:
+                page_meta["image_count"] = len(page_images)
+                page_meta["images"] = _serialize_images(page_images)
+
+            page_meta.update(copy.deepcopy(user_meta))
+            documents.append(Document(content=page_content, meta=page_meta))
+
+        return documents
+
+    def _create_chunked_documents(
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:387]: static method

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py [hunk 26]
```diff
@@ -0,0 +1,518 @@
...
+        # Per-page mode — one Document per page
+        if result.pages:
+            return self._create_per_page_documents(result, base_meta, source_meta, user_meta)
+
+        # Default: unified mode — one Document per source
+        content = self._assemble_content(result.content, result.tables, result.output_format)
+        merged = {**base_meta, **source_meta, **copy.deepcopy(user_meta)}
+        return [Document(content=content, meta=merged)]
+
+    def _create_per_page_documents(
```
Comment (anakin87) [inline integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:355]: could be made static

File: integrations/kreuzberg/tests/test_converter.py [hunk 27]
```diff
@@ -0,0 +1,753 @@
...
+def test_edge_sequential_extraction_error_skipped(
+    mock_extract: MagicMock, sequential_converter: KreuzbergConverter
+) -> None:
+    mock_extract.side_effect = RuntimeError("extraction failed")
+    converter = sequential_converter
+    result = converter.run(sources=[FIXTURES_DIR / "sample.txt"])
+    assert result["documents"] == []
+
+
+def _make_mock_result(**overrides: Any) -> MagicMock:
```
Comment (anakin87) [inline integrations/kreuzberg/tests/test_converter.py]: could we transform this into a fixture?

File: integrations/kreuzberg/tests/test_converter.py (hunk 13 above)
Comment (v-tan) [reply integrations/kreuzberg/tests/test_converter.py:4]: ofcourse, reduced the number by 33%.

Comment (anakin87) [reply integrations/kreuzberg/tests/test_converter.py:4]: thank you

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py (hunk 23 above)
Comment (v-tan) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:526]: batch extraction doesn't raise error but inlines the error in the result for the document.
it returns 
`ExtractionResult(content="Error: ...",  metadata={}, mime_type="text/plain")`

Comment (anakin87) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:526]: I'd align the results of batch and non-batch as much as possible. What do you think?

Comment (v-tan) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:526]: I was just doing that. sending in the fix. corrupt batch results will log warning but not hinder the pipeline, like the sequential flow.

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py (hunk 10 above)
Comment (v-tan) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py]: `ignore_missing_imports = true` didn't work, so use `follow_imports = "skip"`.
I guess mypy could find the impl.

File: integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py (hunk 23 above)
Comment (anakin87) [reply integrations/kreuzberg/src/haystack_integrations/components/converters/kreuzberg/converter.py:526]: thank you

File: integrations/kreuzberg/pyproject.toml [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+cov-retry = 'pytest --cov=haystack_integrations --reruns 3 --reruns-delay 30 -x {args:tests}'
+types = "mypy -p haystack_integrations.components.converters.kreuzberg {args}"
+
+[tool.ruff]
+line-length = 120
+
+[tool.ruff.lint]
+select = [
+    "A",    # flake8-builtins
+    "ARG",  # flake8-unused-arguments
```
Comment (anakin87) [inline integrations/kreuzberg/pyproject.toml:79]:
```suggestion
    "A",    # flake8-builtins
    "ANN",  # flake8-annotations
    "ARG",  # flake8-unused-arguments
```

File: integrations/kreuzberg/pyproject.toml [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+    "RUF",  # ruff-specific rules
+    "S",    # flake8-bandit
+    "T",    # flake8-print
+    "TID",  # flake8-tidy-imports
+    "UP",   # pyupgrade
+    "W",    # pycodestyle (warning)
+    "YTT",  # flake8-2020
+]
+ignore = [
+    # Allow non-abstract empty methods in abstract base classes
```
Comment (anakin87) [inline integrations/kreuzberg/pyproject.toml:107]:
```suggestion
    # Allow Any - used legitimately for dynamic types and SDK boundaries
    "ANN401",
    # Allow non-abstract empty methods in abstract base classes
```

File: integrations/kreuzberg/pyproject.toml [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+
+[tool.ruff.lint.isort]
+known-first-party = ["haystack_integrations"]
+
+[tool.ruff.lint.flake8-tidy-imports]
+ban-relative-imports = "parents"
+
+[tool.ruff.lint.per-file-ignores]
+# Tests can use magic values, assertions, and relative imports
+"tests/**/*" = ["PLR2004", "S101", "S108", "TID252"]
```
Comment (anakin87) [inline integrations/kreuzberg/pyproject.toml]:
```suggestion
"tests/**/*" = ["ANN", "PLR2004", "S101", "S108", "TID252"]
```
